[BE] Remove check_pyobf

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158611
* #158507
* #158503
* __->__ #158493

This pr converts `check_pyobj(bool)` -> `get_pyobj()` as we still need a way of accessing the pyobject however we no longer need to do these checks.

PaliC committed:
[BE] Remove check_pyobf

[ghstack-poisoned]

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -405,17 +405,12 @@ static PyObject* THPModule_swap_tensor_impl(PyObject* _unused, PyObject* args) {
-      "Both tensors should have PyObjects tagged by the current python interpreter");
-  TORCH_CHECK(mb_obj_a.value() == a_);
-  TORCH_CHECK(mb_obj_b.value() == b_);
+  std::optional<PyObject*> obj_a =
```
The return type is not optional anymore.
We should be consistent here to either use optional or the caller has to check for non-null PyObject*

PaliC committed:
Update on "[BE] Remove check_pyobf"


This pr converts `check_pyobj(bool)` -> `get_pyobj()` as we still need a way of accessing the pyobject however we no longer need to do these checks.

[ghstack-poisoned]

albanD commented on c10/core/impl/PyObjectSlot.h:
```diff
@@ -37,36 +37,8 @@ struct C10_API PyObjectSlot {
-    } else {
-      return _unchecked_untagged_pyobj();
-    }
+  PyObject* get_pyobj() const {
```
Can you add some quick doc here on how to check validity of the result here?

albanD commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -585,11 +585,7 @@ static void set_tensor_attr_with_capsule(
-  TORCH_CHECK(
-      mb_obj.has_value(), "Tensor subclass's PyInterpreter has no value");
-  auto obj = mb_obj.value();
+  PyObject* obj = tensor->pyobj_slot()->get_pyobj();
```
You lost the error checking here!

albanD commented on torch/csrc/Storage.cpp:
```diff
@@ -41,48 +41,25 @@ PyObject* THPStorage_NewWithStorage(
-  }
-
-  return obj;
+  auto obj = _storage.unsafeGetStorageImpl()->pyobj_slot()->get_pyobj();
```
NOT ok here for sure, obj could en nullptr

albanD reviewed (changes requested):
I think there is a confusion that obj can be empty ! You need to keep all the alloc code right?

PaliC committed:
Update on "[BE] Remove check_pyobf"


This pr converts `check_pyobj(bool)` -> `get_pyobj()` as we still need a way of accessing the pyobject however we no longer need to do these checks.

[ghstack-poisoned]

PaliC committed:
Update on "[BE] Remove check_pyobf"


This pr converts `check_pyobj(bool)` -> `get_pyobj()` as we still need a way of accessing the pyobject however we no longer need to do these checks.

[ghstack-poisoned]

PaliC committed:
Update on "[BE] Remove check_pyobf"


This pr converts `check_pyobj(bool)` -> `get_pyobj()` as we still need a way of accessing the pyobject however we no longer need to do these checks.

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]
